plugins/reachability/src/main/java/org/qbicc/plugin/reachability/ReachabilityBlockBuilder.java

```diff
@@ -202,150 +199,162 @@ public Value classOf(Value typeId) {
-                processInstantiatedClass(type.getSuperClass(), false, false);
-                info.addLiveClass(type);
+                processInstantiatedClass(type.getSuperClass(), false);
 
-                // For every instance method that is not already enqueued,
-                // check to see if it is overriding an enqueued method and thus should be enqueued.
+                // For every instance method that is not already invokable,
+                // check to see if it is either (a) a deferred invoked method or (b) overriding an invokable method and thus should be enqueued.
                 for (MethodElement im: type.getInstanceMethods()) {
-                    if (!ctxt.wasEnqueued(im)) {
-                        MethodElement overiddenMethod = type.getSuperClass().resolveMethodElementVirtual(im.getName(), im.getDescriptor());
-                        if (overiddenMethod != null && ctxt.wasEnqueued(overiddenMethod)) {
+                    if (!info.isInvokableMethod(im)) {
+                        if (info.isDeferredInstanceMethod(im)) {
+                            rtaLog.debugf("\tnewly reachable class: enqueued deferred instance method: %s", im);
+                            info.removeDeferredInstanceMethod(im);
+                            info.addInvokableMethod(im);
                             ctxt.enqueue(im);
-                            rtaLog.debugf("\tnewly reachable class: enqueued overriding method: %s", im);
+                        } else {
+                            MethodElement overiddenMethod = type.getSuperClass().resolveMethodElementVirtual(im.getName(), im.getDescriptor());
+                            if (overiddenMethod != null && info.isInvokableMethod(overiddenMethod)) {
+                                rtaLog.debugf("\tnewly reachable class: enqueued overriding instance method: %s", im);
+                                info.addInvokableMethod(im);
+                                ctxt.enqueue(im);
+                            }
                         }
                     }
                 }
             }
 
             // Extend the interface hierarchy
             for (LoadedTypeDefinition i: type.getInterfaces()) {
-                processInstantiatedInterface(info, i, false);
+                processInstantiatedInterface(i);
                 info.addInterfaceEdge(type, i);
             }
 
-            // For every enqueued interface method, make sure my implementation of that method is also enqueued.
+            // For every reachable interface method, make sure my implementation of that method is invokable.
             for (LoadedTypeDefinition i: type.getInterfaces()) {
                 for (MethodElement sig: i.getInstanceMethods()) {
-                    if (ctxt.wasEnqueued((sig))) {
+                    if (info.isInvokableMethod(sig)) {
                         MethodElement impl = type.resolveMethodElementVirtual(sig.getName(), sig.getDescriptor());
```

Comment:
Will this catch overrides of interface methods where the interface is only declared on the superclass?  

```
interface I {
  default void m() { .... }
}

class A implements I {}

class B extends A {
  void m() { /* B specific impl */ }
}

Somewhere: new B();

Somewhere else:  I.m();
```

My hesitation about this case comes from `getInterfaces()` only returning the locally defined interfaces

Reply:
It should be caught on the superclass.   ie, the superclass should implement that method, therefore it should have an instance method (real or a synthetic abstract one) that this subclass is either using, or overriding.

Reply:
For at least this fairly simple case, it works as expected.  I wrote two test cases, one where I structured the code so RTA found the `new B()` first and one where it found the `I.m()` first.  Looking at the RTA log output, it handled appropriately.